File: exercise/python/day00/encryption_email/encryption.py
from base64 import b64encode, b64decode

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes


class Encryption:
    def __init__(self, key: str, iv: str):
        self.key = b64decode(key)
        self.iv = b64decode(iv)

    def encrypt(self, plain_text: str) -> str:
        plain_bytes = plain_text.encode('utf-8')
        cipher = Cipher(algorithms.AES(self.key), modes.CBC(self.iv), backend=default_backend())
        encryptor = cipher.encryptor()

        pad_len = 16 - (len(plain_bytes) % 16)
        padded_plaintext = plain_bytes + bytes([pad_len] * pad_len)

        encrypted_bytes = encryptor.update(padded_plaintext) + encryptor.finalize()

        return b64encode(encrypted_bytes).decode('utf-8')

    def decrypt(self, encrypted_text: str) -> str:
        encrypted_bytes = b64decode(encrypted_text)

        cipher = Cipher(algorithms.AES(self.key), modes.CBC(self.iv), backend=default_backend())
        decryptor = cipher.decryptor()

        decrypted_padded_bytes = decryptor.update(encrypted_bytes) + decryptor.finalize()
        pad_len = decrypted_padded_bytes[-1]
        decrypted_bytes = decrypted_padded_bytes[:-pad_len]

        return decrypted_bytes.decode('utf-8')

File: exercise/ts/day00/src/encryption.ts
import * as crypto from 'crypto';

export class Encryption {
    private readonly key: Buffer;
    private readonly iv: Buffer;

    constructor(key: string, iv: string) {
        this.key = Buffer.from(key, 'base64');
        this.iv = Buffer.from(iv, 'base64');
    }

    encrypt(plainText: string): string {
        const cipher = crypto.createCipheriv('aes-256-cbc', this.key, this.iv);
        let encrypted = cipher.update(plainText, 'utf-8', 'base64');
        encrypted += cipher.final('base64');

        return encrypted;
    }

    decrypt(encryptedText: string): string {
        const decipher = crypto.createDecipheriv('aes-256-cbc', this.key, this.iv);
        let decrypted = decipher.update(encryptedText, 'base64', 'utf-8');
        decrypted += decipher.final('utf-8');

        return decrypted;
    }
}